src/thread.rs

```diff
@@ -120,22 +196,51 @@ where
     // Publish the result and unblock the waiter. We need to do this now, because once this
     // closure completes, the Execution will consider this task Finished and invoke the
     // scheduler.
-    *result.lock().unwrap() = Some(Ok(ret));
+    *result.borrow_mut() = Some(Ok(ret));
     ExecutionState::with(|state| {
         if let Some(waiter) = state.current_mut().take_waiter() {
             state.get_mut(waiter).unblock();
         }
     });
 }
 
+/// An owned permission to join on a scoped thread (block on its termination).
+///
+/// See [`Scope::spawn`] for details.
+#[derive(Debug)]
+pub struct ScopedJoinHandle<'scope, T> {
+    handle: JoinHandle<T>,
+    _marker: PhantomData<&'scope T>,
+}
+
+#[derive(Clone, Debug)]
+struct ScopeData {
+    num_running_threads: Rc<RefCell<usize>>,
+    main_task: TaskId,
+}
+
+impl<T> ScopedJoinHandle<'_, T> {
+    /// Waits for the associated thread to finish.
+    pub fn join(self) -> Result<T> {
+        self.handle.join()
+    }
+
+    /// Extracts a handle to the underlying thread.
+    pub fn thread(&self) -> &Thread {
+        self.handle.thread()
+    }
+}
+
 /// An owned permission to join on a thread (block on its termination).
 #[derive(Debug)]
 pub struct JoinHandle<T> {
     task_id: TaskId,
     thread: Thread,
-    result: std::sync::Arc<std::sync::Mutex<Option<Result<T>>>>,
+    result: Rc<RefCell<Option<Result<T>>>>,
 }
 
+unsafe impl<T: Send> Send for JoinHandle<T> {}
```

Comment:
`JoinHandle` is `Send` and `Sync` (and is so regardless of `T`). I see now that we currently have done this incorrectly, put up a PR: https://github.com/awslabs/shuttle/pull/183.